### PR TITLE
[Backport 3.3.x] [Fixes #9013] Advanced Workflow: workflow logic should take precedence over standard permissions

### LINKED ISSUE #9013: Advanced Workflow: workflow logic should take precedence over standard permissions
Currently there's still some mixed logic affecting the Advanced Workflow leading to possible inconsitent states.

e.g.:
 1. When removing the `owner` from the current `GroupProfile`, the `managers` and `members` are not updated accordingly
 2. When setting/removing the `GroupProfile` from the resource `metadata`, the `managers` and `members` are not updated accordingly
 3. When promoting/demoting a `member` of the `GroupProfile`, the permissions are not updated accordingly

What we would like to achieve, first of all, is to fix the Advanced Workflow permissions accordingly to the following schema:

  | RESOURCE_PUBLISHING | ADMIN_MODERATE_UPLOADS |
-- | -- | -- |
AUTO PUBLISH | X | X |
SIMPLE PUBLISHING | V | X |
SIMPLIFIED WORKFLOW | X | V |
ADVANCED WORKFLOW | V | V |

  | New Resource Uploaded | Resource Approved | Resource Published | Manager Demoted | Member Promoted | Member Removed
-- | -- | -- |-- | -- |-- | -- | 
AUTO PUBLISH | - is_approved = True <br> - is_published = True <br> - OWNER = {ADMIN} | - | - | - | - | -
SIMPLE PUBLISHING | - is_approved = True <br> - is_published = False <br> - OWNER = {"publish": False} <br> - MANAGER = {"publish": True} <br> - MEMBER = {"view", "download"} <br> - ANONYMOUS = {} | - | - is_approved = True <br> - is_published = True <br> - OWNER = {"publish": False} <br> - MANAGER = {"publish": True} <br> - MEMBER = {"view", "download"} <br> - ANONYMOUS = {"view", "download"} | - {"view", "download"} | - {ALL + "publish": True} | - {"view", "download"}
SIMPLIFIED WORKFLOW | - is_approved = False <br> - is_published = True <br> - OWNER = {"publish": False} <br> - MANAGER = {"publish": True} <br> - MEMBER = {"view", "download"} <br> - ANONYMOUS = {} | - is_approved = True <br> - is_published = True <br> - OWNER = {"publish": False, "change_resourcebase": False, "change_resource_metadata": False} <br> - MANAGER = {"publish": True} <br> - MEMBER = {"view", "download"} <br> - ANONYMOUS = {"view", "download"} | - | - {"view", "download"} | - {ALL + "publish": True} | - {"view", "download"}
ADVANCED WORKFLOW | - is_approved = False <br> - is_published = False <br> - OWNER = {"publish": False, "change_permissions": False} <br> - MANAGER = {"publish": True} <br> - MEMBER = {"view", "download"} <br> - ANONYMOUS = {} | - is_approved = True <br> - is_published = False <br> - OWNER = {"publish": False, "change_permissions": False, "change_resourcebase": False, "change_resource_metadata": False} <br> - MANAGER = {"publish": True} <br> - MEMBER = {"view", "download"} <br> - ANONYMOUS = {} | - is_approved = True <br> - is_published = True <br> - OWNER = {"publish": False, "change_permissions": False, "change_resourcebase": False, "change_resource_metadata": False} <br> - MANAGER = {"publish": True} <br> - MEMBER = {"view", "download"} <br> - ANONYMOUS = {"view", "download"} | - {"view", "download"} | - {ALL + "publish": True} | - {"view", "download"}

### AUTO PUBLISH
* The `OWNER` has full permissions on the resource; he can always EDIT/DELETE/APPROVE/PUBLISH
* A `MANAGER` has full permissions on the resource; he can always EDIT/DELETE/APPROVE/PUBLISH
* A `MEMBER` can VIEW/DOWNLOAD
* `ANONYMOUS` can VIEW/DOWNLOAD if the resource is `PUBLISHED`

_If the OWNER is also a MANAGER, the MANAGER permissions take always precedence_

#### Approve
* `ANONYMOUS` can VIEW/DOWNLOAD

#### Publish
* `ANONYMOUS` can VIEW/DOWNLOAD

#### Group set/remove
* Updates the `MANAGER`/`MEMBER` accordingly

#### Remove a `MEMBER`
* Removes the permissions accordingly

### SIMPLE PUBLISHING
* The `OWNER` can always EDIT/DELETE
* A `MANAGER` can always EDIT/DELETE/APPROVE/PUBLISH
* A `MEMBER` can VIEW/DOWNLOAD
* `ANONYMOUS` can VIEW/DOWNLOAD if the resource is `PUBLISHED`

_If the OWNER is also a MANAGER, the MANAGER permissions take always precedence_

#### Approve
* No changes

#### Publish
* `ANONYMOUS` can VIEW/DOWNLOAD

#### Group set/remove
* Updates the `MANAGER`/`MEMBER` accordingly

#### Remove a `MEMBER`
* Removes the permissions accordingly

### SIMPLIFIED WORKFLOW
* The `OWNER` can EDIT/DELETE if the resource is **not** `APPROVED` or `PUBLISHED`
* A `MANAGER` can always EDIT/DELETE/APPROVE/PUBLISH
* A `MEMBER` can VIEW/DOWNLOAD
* `ANONYMOUS` can VIEW/DOWNLOAD if the resource is `APPROVED` or `PUBLISHED`

_If the OWNER is also a MANAGER, the MANAGER permissions take always precedence_

#### Approve
* The `OWNER` cannot EDIT/DELETE anymore

#### Publish
* The `OWNER` cannot EDIT/DELETE anymore
* `ANONYMOUS` can VIEW/DOWNLOAD

#### Group set/remove
* Updates the `MANAGER`/`MEMBER` accordingly

#### Remove a `MEMBER`
* Removes the permissions accordingly

### ADVANCED WORKFLOW
* The `OWNER` can EDIT/DELETE if the resource is **not** `APPROVED` or `PUBLISHED`
* The `OWNER` **cannot** `CHANGE_PERMISSIONS`
* A `MANAGER` can always EDIT/DELETE/APPROVE/PUBLISH
* A `MEMBER` can VIEW/DOWNLOAD
* `ANONYMOUS` can VIEW/DOWNLOAD if the resource is `APPROVED` or `PUBLISHED`

_If the OWNER is also a MANAGER, the MANAGER permissions take always precedence_

#### Approve
* The `OWNER` cannot EDIT/DELETE anymore

#### Publish
* The `OWNER` cannot EDIT/DELETE anymore
* `ANONYMOUS` can VIEW/DOWNLOAD

#### Group set/remove
* Updates the `MANAGER`/`MEMBER` accordingly

#### Remove a `MEMBER`
* Removes the permissions accordingly

### PR DESCRIPTION
**Pull request by afabiani**
References: #9013

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [x] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

**Comment by marthamareal** [comment]
@afabiani , On testing

For `ADVANCED WORKFLOW`
    - If the `owner group` is different from the `resource group`, The group manager of the resource group is unable to access the resource.
    - 
For `SIMPLIFIED WORKFLOW`
    - Resource is created when both `is_published and is_approved` = False
    - Approving the resource has no effect on owner perms and also Anonymous can not access it.

For `SIMPLE PUBLISHING`
    - Resource is created when both `is_published and is_approved` = False

**Comment by marthamareal** [comment]
@afabiani

1. For `SIMPLIFIED WORKFLOW` Anonymous can not see a resource after updating the fields `is_published and is_approved` to `True`.

2. An example for `ADVANCED WORKFLOW` issue:
- create a resource as owner in group 1
- Login as an admin and assign the above resource to group 2
- Notice Managers of group2 can not access the above resource

**Review by mattiagiupponi — CHANGES_REQUESTED** [review]
@afabiani which are the part related to the advanced workflow? there are a lot of different development here (OGC_Servers_Handler, test changed, new tests) which looks like that are out of the issue scope

**Comment by afabiani** [comment]
@marthamareal please check again, the issues you pointed out should be new fixed.

File: geonode/base/models.py [hunk 1]
```diff
@@ -1096,6 +1097,9 @@ def save(self, notify=False, *args, **kwargs):
                 recipients = get_notification_recipients(notice_type_label, resource=self)
                 send_notification(recipients, notice_type_label, {'resource': self})
             elif self.pk:
+                # Group has changed
+                _group_status_changed = self.group != ResourceBase.objects.get(pk=self.get_self_resource().pk).group
```
Comment (mattiagiupponi) [inline geonode/base/models.py:1101]: If you already have the resource, is not enough to do:
```python
_group_status_changed = self.group != self.get_self_resource().group
```

**Pull request merged**